sklearn/neighbors/_lof.py

```diff
@@ -462,7 +468,15 @@ def score_samples(self, X):
         distances_X, neighbors_indices_X = self.kneighbors(
             X, n_neighbors=self.n_neighbors_
         )
-        X_lrd = self._local_reachability_density(distances_X, neighbors_indices_X)
+
+        # To preserve inputs dtype for outputs.
+        # This might downcast distances_X from np.float64 to np.float32.
+        distances_X = distances_X.astype(X.dtype, copy=False)
```

Comment:
Since `dtype="numeric"` in `check_array`, `X.dtype` will preserve and `distance_X` can be cast to whatever numerical `dtype`.

For `distances_X`, I don't think we want to allow to cast to integral `dtype`.

We probably want to have a safe guard here:

```python
if X.dtype == np.float32:
    distances_X = distances_X.atype(X.dtype, copy=False)
```